### ISSUE TITLE
reason behind GAIN_MIX_MAX hardcoded value of 16?

### ISSUE THREAD

**Issue by Boris-Barboris** [kcat/openal-soft#252]
Is there a particular reason why GAIN_MIX_MAX is hardcoded to 16.0 here?
https://github.com/kcat/openal-soft/blob/31b54eb86a25457e561f4e27324163def0a8bd66/OpenAL32/Include/alu.h#L296

Can it be larger? Is it clamped by backend capability as well?

**Comment by kcat** [comment]
It's like that to avoid excessive gain increase, since 1.0 (+0dB) represents the normal unattenuated volume for sounds. Depending on the sound, increasing a source's gain above 1 can make samples exceed 0dbFS (the normalized range), and it can either have no effect or introduce clipping distortion. Is there an issue you have where you need it increased even more?

**Comment by Boris-Barboris** [comment]
Yeah, in my case I stream 1-second 16bit pcb chunks from the server to client that have a very high dynamic range, and provide the user with a volume slider. Amplitude in quiet state is around 100 (out of 2^15), so I would need a max gain of around 320 (+50 dB) to saturate the output.

**Comment by kcat** [comment]
> Amplitude in quiet state is around 100 (out of 2^15), so I would need a max gain of around 320 (+50 dB) to amplify the output to saturation.

With a +50dB increase for a 16-bit stream, that would result in quality worse than 8-bit (about 5 or 6 bits, by my rough estimation). For samples that exceed -50dBFS, some systems would get horrendous clipping, perhaps even damaging speakers or causing pain for listeners (especially on headphones).

A "quiet state" and "saturation [of the output]" are also antithetical. If you want to utilize the dynamic range available, quiet sounds should stay quiet so they stand in contrast to loud sounds. If you want all sounds to be loud and saturate the output regardless if they're normally quiet or loud, you probably want a compressor to reduce the dynamic range (which would ideally be applied before 16-bit quantization, to reduce the quantization noise).

**Comment by Boris-Barboris** [comment]
Everything you say is true, but please just assume I know exactly what I want. Nor bad quality, nor possible clipping are a problem in my case, they are expected and allowed. I do not want all sound to be loud, it's magnitude has a physical meaning (hydroacoustics simulation), hence I am not interested in a compressor.

If you don't want to bother, that's all right, I'll stay in the fork.
If you want me to PR, please advise me on the implementation details - new constant or parameter.

**Comment by kcat** [comment]
I'm not saying no, I'm just trying to figure out what's needed, and whether something else may do the job. From what you say, that doesn't seem to be the case.

The only technical thing it needs to prevent is samples in the final mix becoming +/-infinity, which is sticky (there's very few values you can add, subtract, multiply, or divide by to get non-infinity; worse, multiplying infinity by 0 results in NaN). That obviously leaves a lot of room that the max gain can be increased by, given a limited number of sources. Other than that, it's just about what's practical. 16, 320, 16384... It could also be made so that the increased max gain is only allowed if OpenAL Soft's limiter is enabled, to ensure no excessive clipping (allow 4 or 8 or something by default, and if the limiter is enabled allow 16384).
